scripts/target.js

```diff
@@ -0,0 +1,281 @@
+        if (section.getAttribute('data-section-status') === 'loaded') {
+          // eslint-disable-next-line no-console
+          console.debug('section has been loaded... resolving', section);
+          observer.disconnect();
+          resolve(section);
+        }
+      });
+      observer.observe(section, config);
+    }));
+}
+
+/**
+ * Render offers in a section.
+ * @param section The section.
+ * @param offers The offers.
+ */
+function displayOffers(section, offers) {
+  offers.forEach((offer) => {
+    const { type, cssSelector, content } = offer;
+    const targetElement = section.querySelector(escapeSelector(cssSelector));
+    if (targetElement) {
+      switch (type) {
+        case 'insertAfter':
+          console.debug('inserting content after', targetElement); // eslint-disable-line no-console
+          targetElement.insertAdjacentHTML('afterend', content);
+          break;
+        case 'insertBefore':
+          console.debug('inserting content before', targetElement); // eslint-disable-line no-console
+          targetElement.insertAdjacentHTML('beforebegin', content);
+          break;
+        case 'setHtml':
+          console.debug('setting content of', targetElement); // eslint-disable-line no-console
+          targetElement.innerHTML = content;
+          break;
+        default:
+          console.warn(`unsupported action type ${type}`); // eslint-disable-line no-console
+      }
+      console.debug('section has been rendered', section); // eslint-disable-line no-console
+      window?.measurePerformance(
+        `targeting:rendering-section:${Array.from(section.classList).join('_')}`,
+      );
+    }
+  });
+  if (section.style.visibility === 'hidden') {
+    // eslint-disable-next-line no-console
+    console.debug('revealing section', section);
+    section.style.visibility = 'visible';
+  }
+}
+
+/**
+ * Get the section for a selector.
+ * @param selector The element selector.
+ */
+function getSectionByElementSelector(selector) {
+  let section = document.querySelector(escapeSelector(selector));
+  while (section && !section.classList.contains('section')) {
+    section = section.parentNode;
+  }
+  return section;
+}
+
+/**
+ * Start targeting for a client on a host.
+ * @param client The client.
+ * @param useProxy Whether to use the proxy.
+ */
+export default function loadOffers(client, useProxy) {
+  if (window.location.href.includes('adobe_authoring_enabled')) {
+    // eslint-disable-next-line no-console
+    console.debug('authoring enabled... skipping targeting');
+    return;
+  }
+
+  window?.createPerformanceMark('targeting:loading-offers');
+
+  const sessionId = getSessionId();
+  // eslint-disable-next-line no-console
+  console.debug(`Using session ID ${sessionId}`);
+
+  document.body.style.visibility = 'hidden';
+
+  const pendingOffers = fetchOffers(client, sessionId, useProxy ?? window.location.host.endsWith('workers.dev'));
+
+  getDecoratedContent()
```

Comment:
I don't think it makes sense to wait for the `appear` class on the body. This is set after the main decoration finished, but before any block got decorated. In that state all sections are hidden. We need to apply the offers when the sections appear. So this is not necessary.

Reply:
Good point. I could hide the entire body part instead of just the main, which has to be awaited as it initially does not exist. Why do we hide the main?

The problem is that fetching offers for a webpage can take longer than expected. If the offers arrive late, after the main content of the page (LCP) has loaded, it causes issues like flickering and showing default content. To prevent this, we hide the main part of the page as a precaution, because we don't know yet which specific sections the offers will target.

@ramboz @Buuhuu thoughts?